File: leetcode/find-the-town-judge.py
# 997. Find the Town Judge
# 🟢 Easy
#
# https://leetcode.com/problems/find-the-town-judge/
#
# Tags: Array - Hash Table - Graph

import timeit
from typing import List


# We can solve this problem using topological sorting, create sets of
# elements that are trusted by the current one and elements that the
# current one trusts, return the element that does not trust any others
# and is trusted by n-1 others.
#
# Time complexity: O(n+t) - Where n is n and t is the number of items
# in the trust array. We iterate over all the elements in trust to
# create the topological sorting, then iterate over a max of n elements
# to find the one that matches the given conditions.
#
# Runtime 736 ms Beats 88.56%
# Memory 18.9 MB Beats 59.7%
class UseSets:
    def findJudge(self, n: int, trust: List[List[int]]) -> int:
        # In and outdegree arrays.
        trusted_by, trusts = [set() for _ in range(n + 1)], [
            set() for _ in range(n + 1)
        ]
        for trustee, trusted in trust:
            trusted_by[trusted].add(trustee)
            trusts[trustee].add(trusted)
        for i in range(1, n + 1):
            if len(trusted_by[i]) == n - 1 and len(trusts[i]) == 0:
                return i
        # We can't determine the judge.
        return -1


# We can solve this problem using topological sorting, count the
# elements that are trusted by the current one and elements that the
# current one trusts, return the element that does not trust any others
# and is trusted by n-1 others.
#
# Time complexity: O(n+t) - Where n is n and t is the number of items
# in the trust array. We iterate over all the elements in trust to
# create the topological sorting, then iterate over a max of n elements
# to find the one that matches the given conditions.
#
# Runtime 726 ms Beats 94.15%
# Memory 18.9 MB Beats 59.7%
class Count:
    def findJudge(self, n: int, trust: List[List[int]]) -> int:
        # In and outdegree arrays.
        trusted_by, trusts = [0] * (n + 1), [0] * (n + 1)
        for trustee, trusted in trust:
            trusted_by[trusted] += 1
            trusts[trustee] += 1
        for i in range(1, n + 1):
            if trusted_by[i] == n - 1 and trusts[i] == 0:
                return i
        # We can't determine the judge.
        return -1


def test():
    executors = [
        UseSets,
        Count,
    ]
    tests = [
        [1, [], 1],
        [2, [[1, 2]], 2],
        [3, [[1, 3], [2, 3]], 3],
        [3, [[1, 3], [2, 3], [3, 1]], -1],
    ]
    for executor in executors:
        start = timeit.default_timer()
        for _ in range(1):
            for col, t in enumerate(tests):
                sol = executor()
                result = sol.findJudge(t[0], t[1])
                exp = t[2]
                assert result == exp, (
                    f"\033[93m» {result} <> {exp}\033[91m for"
                    + f" test {col} using \033[1m{executor.__name__}"
                )
        stop = timeit.default_timer()
        used = str(round(stop - start, 5))
        cols = "{0:20}{1:10}{2:10}"
        res = cols.format(executor.__name__, used, "seconds")
        print(f"\033[92m» {res}\033[0m")


test()

File: leetcode/find-the-town-judge.rs
// 997. Find the Town Judge
// 🟢 Easy
//
// https://leetcode.com/problems/find-the-town-judge/
//
// Tags: Array - Hash Table - Graph

struct Solution;
impl Solution {
    /// For each entry, count the number of people that trust it and the number of people it trust,
    /// we can use one vector because the amount of people that trust the town judge minus the
    /// amount of people they trust is n-1.
    ///
    /// Time complexity: O(n+t) - We iterate over the t trust relations to get the counts, then we
    /// iterate over the n counts to get the result. In terms of a graph this would be v+e.
    /// Space complexity: O(n) - We store the results in one vector of size n.
    ///
    /// Runtime 14 ms Beats 100%
    /// Memory 2.74 MB Beats 78.26%
    pub fn find_judge(n: i32, trust: Vec<Vec<i32>>) -> i32 {
        let m = n as usize + 1;
        let mut counter = vec![0; m];
        for t in trust {
            counter[t[0] as usize] -= 1;
            counter[t[1] as usize] += 1;
        }
        for i in 1..m {
            if counter[i] == n - 1 {
                return i as i32;
            }
        }
        -1
    }

    // Same logic as the previous solution, but using two vectors, one for each, this solution
    // would be more adaptable to other conditions, like "the judge trusts a maximum of x people".
    //
    /// Time complexity: O(n+t) - We iterate over the t trust relations to get the counts, then we
    /// iterate over the n counts to get the result. In terms of a graph this would be v+e.
    /// Space complexity: O(n) - We store the results in one vector of size n.
    //
    // Runtime 14 ms Beats 100%
    // Memory 2.8 MB Beats 36%
    #[allow(dead_code)]
    pub fn find_judge_two_vectors(n: i32, trust: Vec<Vec<i32>>) -> i32 {
        let mut trusted_by = vec![0; (n as usize) + 1];
        let mut trusts = vec![0; (n as usize) + 1];
        for t in trust {
            trusts[t[0] as usize] += 1;
            trusted_by[t[1] as usize] += 1;
        }
        for i in 1..=(n as usize) {
            if trusted_by[i] == (n as usize) - 1 && trusts[i] == 0 {
                return i as i32;
            }
        }
        -1
    }
}

// Tests.
fn main() {
    let tests = [
        (1, vec![], 1),
        (2, vec![[1, 2]], 2),
        (3, vec![[1, 3], [2, 3]], 3),
        (3, vec![[1, 3], [2, 3], [3, 1]], -1),
    ];
    println!("\n\x1b[92m» Running {} tests...\x1b[0m", tests.len());
    let mut success = 0;
    for (i, t) in tests.iter().enumerate() {
        let res = Solution::find_judge(t.0, t.1.clone().iter().map(|arr| arr.to_vec()).collect());
        if res == t.2 {
            success += 1;
            println!("\x1b[92m✔\x1b[95m Test {} passed!\x1b[0m", i);
        } else {
            println!(
                "\x1b[31mx\x1b[95m Test {} failed expected: {:?} but got {}!!\x1b[0m",
                i, t.2, res
            );
        }
    }
    println!();
    if success == tests.len() {
        println!("\x1b[30;42m✔ All tests passed!\x1b[0m")
    } else if success == 0 {
        println!("\x1b[31mx \x1b[41;37mAll tests failed!\x1b[0m")
    } else {
        println!(
            "\x1b[31mx\x1b[95m {} tests failed!\x1b[0m",
            tests.len() - success
        )
    }
}
